File: data-structures/queue/go/queue.go
package main

import (
	"fmt"
)

type QueueNode struct {
	data int
	next *QueueNode
}

type Queue struct {
	head *QueueNode
	size int
}

func (q *Queue) Enqueue(data int) {
	newNode := &QueueNode{data: data}
	if q.head == nil {
		q.head = newNode
		q.size++
		return
	}

	curr := q.head
	for curr.next != nil {
		curr = curr.next
	}
	curr.next = newNode
	q.size++
}

func (q *Queue) Dequeue() int {
	if q.head == nil {
		fmt.Println("Queue is empty")
		return -1
	}
	val := q.head.data
	q.head = q.head.next
	q.size--
	return val
}

func (q *Queue) Front() int {
	if q.head == nil {
		fmt.Println("Queue is empty")
		return -1
	}
	return q.head.data
}

func (q *Queue) Size() int {
	return q.size
}

func (q *Queue) IsEmpty() bool {
	return q.size == 0
}

func main() {
	q := &Queue{}

	q.Enqueue(1)
	q.Enqueue(64)
	q.Enqueue(23)
	q.Enqueue(6)

	fmt.Println("Front:", q.Front())
	fmt.Println("Size:", q.Size())

	q.Dequeue()
	fmt.Println("Front:", q.Front())
	fmt.Println("Size:", q.Size())

	fmt.Println("Dequeue:", q.Dequeue()) // 64
	fmt.Println("Dequeue:", q.Dequeue()) // 23
	fmt.Println("Dequeue:", q.Dequeue()) // 6
	fmt.Println("Dequeue:", q.Dequeue()) // Queue is empty → -1
}

File: data-structures/queue/ts/Queue.ts
/**
 * @file Queue.ts
 * @description Provides Userdefined queue
 * @author JestiferHarold
 * @created 2025-07-15
 * 
 * This File Includes:
 * - Queue
 */

import { NewNode } from "../../Node/Node";
//@ts-ignore
import { process } from "node:process";

/**
 * Represents an abstract data type Queue using linked list 
 */
class Queue {
    
    /**
     * Pointer to the first node in the queue
     * @memberof Queue
     */
    private front: NewNode | null;

    /**
     * Initializes the Class Queue
     * @param front NewNode object that is to be the first element in the queue or null, defaulted to null
     */
    constructor (front: NewNode | null = null) {
        this.front = front;
    }

    /**
     * returns the first node without removing 
     * @returns NewNode object if the queue is not null, else null
     */
    getFront(): NewNode | null {
        return this.front;
    }

    /**
     * addes an node object at the end of the queue
     * @param node NewNode object to be added at the end of the queue
     */
    enQueue(node: NewNode): void {
        let n: NewNode | null = this.front;
        for (; n?.getNextNode() != null; n = n.getNextNode());
        node.setNextNode(null);
        n?.setNextNode(node);
    }

    /**
     * removes and returns the first element in the queue
     * @returns the front node object otherwise null
     */
    deQueue(): NewNode | null {
        if (this.front == null) {
            return null;
        }
        //@ts-ignore
        let n: NewNode = this.front;
        this.front = this.front.getNextNode();
        return n;
    }

    /**
     * counts the number of elements in the queue
     * @returns number of nodes present in the queue
     */
    size(): number {
        let i: number = 0;
        for (let n: NewNode | null = this.front; n != null; n = n.getNextNode(), i++);
        return i;
    }

    /**
     * checks if the queue is empty
     * @returns true if the front element is empty, false
     */
    isEmpty(): boolean {
        return this.front == null;
    }

    /**
     * prints the queue in the console
     */
    display(): void {
        let n: NewNode | null = this.front;
        for (; n?.getNextNode() != null; n = n.getNextNode())
            //@ts-ignore
            process.std.write(`${n.getValue()} <- `);
        //@ts-ignore
        process.std.write();
    }

}

export {
    Queue
} 